samza-core/src/main/java/org/apache/samza/coordinator/InputStreamRegexMonitor.java

```diff
@@ -0,0 +1,242 @@
+
+    log.info("Created {} with inputRegexMonitorPeriodMs: {} and systemRegexesToMonitor: {}", this.getClass().getName(),
+        this.inputRegexMonitorPeriodMs, this.systemRegexesToMonitor);
+  }
+
+  /**
+   * Starts the monitor.
+   */
+  public void start() {
+    synchronized (lock) {
+      switch (state) {
+        case INIT:
+          if (inputRegexMonitorPeriodMs > 0) {
+            schedulerService.scheduleAtFixedRate(new Runnable() {
+              @Override
+              public void run() {
+                monitorInputRegexes();
+              }
+            }, 0, inputRegexMonitorPeriodMs, TimeUnit.MILLISECONDS);
+          }
+          state = State.RUNNING;
+          break;
+
+        case RUNNING:
+          // start is idempotent
+          return;
+
+        case STOPPED:
+          throw new IllegalStateException("InputStreamRegexMonitor was stopped and cannot be restarted.");
+      }
+    }
+  }
+
+  /**
+   * Stops the monitor. Once it stops, it cannot be restarted.
+   */
+  public void stop() {
+    synchronized (lock) {
+      // We could also wait for full termination of the scheduler service, but it is overkill for
+      // our use case.
+      schedulerService.shutdownNow();
+
+      state = State.STOPPED;
+    }
+  }
+
+  private void monitorInputRegexes() {
+    log.debug("Running monitorInputRegexes");
+
+    try {
+      // obtain the list of SysStreams that match given patterns for all systems
+      Set<SystemStream> inputStreamsMatchingPattern = new HashSet<>();
+
+      // For each input system, for which we have a regex to monitor
+      for (String systemName : this.systemRegexesToMonitor.keySet()) {
+
+        try {
+          // obtain the list of SysStreams that match the regex for this system
+          // using the systemAdmin in the metadataCache
+          inputStreamsMatchingPattern.addAll(
+              JavaConverters.setAsJavaSetConverter(this.metadataCache.getAllSystemStreams(systemName))
+                  .asJava()
+                  .stream()
+                  .filter(x -> x.getStream().matches(this.systemRegexesToMonitor.get(systemName).pattern()))
+                  .collect(Collectors.toSet()));
+
+        } catch (UnsupportedOperationException e) {
+          log.error("UnsupportedOperationException while monitoring input regexes for system {}", systemName, e);
+        }
+      }
+
+      // if there is a stream that is in the input-Set but not in the streamsToMonitor
+      // since streamsToMonitor = task.inputs
+      if (!streamsToMonitor.containsAll(inputStreamsMatchingPattern)) {
+        log.info("New input system-streams discovered. InputStreamsMatchingPattern: {} but streamsToMonitor: {} ",
+            inputStreamsMatchingPattern, streamsToMonitor);
+
+        // invoke notify callback with new input streams
+        this.callbackMethod.onNewInputStreamsDiscovered(Sets.difference(inputStreamsMatchingPattern, streamsToMonitor));
```

Comment:
probably useful to have "initial set of streams", "regex", "new set of streams" in the callback?